wordpress.org/public_html/wp-content/plugins/plugin-directory/class-plugin-release.php

```diff
@@ -0,0 +1,420 @@
+		}
+
+		$parent_plugin = get_post( $release_post->post_parent );
+		if ( ! $parent_plugin || 'plugin' !== $parent_plugin->post_type ) {
+			return new \WP_Error( 'invalid_plugin', 'Invalid plugin' );
+		}
+
+		$post_status = ( 'trunk' === $release['tag'] ) ? 'draft' : 'publish';
+		$post_title = ( 'trunk' === $release['tag'] ) ? 'trunk' : $release['version'];
+
+		$release_id = wp_update_post( array(
+			'ID'           => $release_post->ID,
+			'post_type'   => 'plugin_release',
+			'post_title'  => $post_title,
+			'post_name'   => $parent_plugin->post_name . '-' . $release['version'],
+			'post_parent' => $parent_plugin->ID,
+			'post_status' => $post_status,
+			'post_date'   => $release_date, // And/or post_date_gmt?
+			// Mirrors the metadata.
+			'meta_input'  => array(
+				'release_date'      => $release['date'],
+				'release_tag'       => $release['tag'],
+				'release_version'   => $release['version'],
+				'release_committer' => $release['committer'],
+				'release_zips_built' => $release['zips_built'],
+				'release_confirmations_required' => $release['confirmations_required'],
+				'release_revision' => $release['revision'],
+				'release_commit_log' => $release['commit_log'] ?? null,
+				'release_tested' => $release['tested'] ?? null,
+				'release_requires_php' => $release['requires_php'] ?? null,
+				'release_requires_wp' => $release['requires_wp'] ?? null,
+				'release_requires_plugins' => $release['requires_plugins'] ?? null,
+			),
+			// TODO: what else? Could store the changelog or other content at the point of release for comparison purposes.
+		) );
+
+		return $release_id;
+	}
+
+	/**
+	 * Save draft (trunk) release for a plugin.
+	 */
+	public function add_or_update_draft_release( $plugin, $release ) {
+		$plugin = get_post( $plugin );
+
+		// Tag must be 'trunk' for this to be a draft release.
+		if ( 'trunk' !== $release['tag'] ) {
+			return new \WP_Error( 'invalid_tag', 'Invalid tag' );
+		}
+
+		if ( !$plugin || 'plugin' !== $plugin->post_type ) {
+			return new \WP_Error( 'invalid_plugin', 'Invalid plugin' );
+		}
+
+		// If there's already a published release for this plugin, we only create a draft if there are unreleased trunk commits.
+		$last_release = $this->get_release( $plugin, null );
+		if ( $last_release ) {
+			if ( !empty( $release[ 'revision' ] ) ) {
+				// Don't create a draft unless the revision number is higher than the last release.
+				if ( max( $release['revision'] ) <= max( $last_release->release_revision ) ) {
+					return false; // Not an error, just skip.
+				}
+			} else {
+				// If we don't have revision numbers, use dates. Maybe this should be removed.
+				if ( strtotime( $release['date'] ) <= strtotime( $last_release->release_date ) ) {
+					return false; // Not an error, just skip.
+				}
+			}
+		}
+
+		// Store the commit log in postmeta. We'll only do this for drafts.
+		$last_release_revision = max( $last_release->release_revision );
```

Comment:
@tellyworth I think this block needs to be inside of the `if ( $last_release )`. I just ran this with a new plugin and this errored out.

Reply:
My assumption is incorrect. First time drafts won't ever receive commits.

Reply:
Okay, so it looks like we could pass `SVN::log( $trunk_url, [ 1, 'HEAD' ] )` to get all the commits for new plugins without releases. 

However, what happens when we roll this out for old plugins that used trunk. We would pull in the whole plugin commit log. Do we care if we show something like that denotes 100 commits:

<img width="992" alt="Screenshot 2024-12-11 at 3 55 31 PM" src="https://github.com/user-attachments/assets/9ea28d5d-0269-4612-88d0-129db5680231">

To those users those commits are not new, however, it could be argued that they are new to the tag/release.

Reply:
Yeah [1, 'HEAD'] will probably work for the initial version or where no release exists.

How about if I limit the commit log meta to the most recent 5 or 10 items for each release?

Reply:
I think that would be less expected and maybe more confusing...? I think we maybe just list them all and see what users say?